server/application/application.go

```diff
@@ -102,47 +102,46 @@ func (s *Server) Update(ctx context.Context, a *appv1.Application) (*appv1.Appli
 
 // UpdateSpec updates an application spec
 func (s *Server) UpdateSpec(ctx context.Context, q *ApplicationSpecRequest) (*appv1.ApplicationSpec, error) {
-	err := s.validateApp(ctx, q.Spec)
+	err := s.validateApp(ctx, &q.Spec)
 	if err != nil {
 		return nil, err
 	}
 	patch, err := json.Marshal(map[string]appv1.ApplicationSpec{
-		"spec": *q.Spec,
+		"spec": q.Spec,
 	})
 	if err != nil {
 		return nil, err
 	}
-	_, err = s.appclientset.ArgoprojV1alpha1().Applications(s.ns).Patch(q.AppName, types.MergePatchType, patch)
-	return q.Spec, err
+	_, err = s.appclientset.ArgoprojV1alpha1().Applications(s.ns).Patch(*q.AppName, types.MergePatchType, patch)
+	return &q.Spec, err
 }
 
 // Delete removes an application and all associated resources
 func (s *Server) Delete(ctx context.Context, q *DeleteApplicationRequest) (*ApplicationResponse, error) {
-	var err error
-	server := q.Server
-	namespace := q.Namespace
-	if server == "" || namespace == "" {
-		server, namespace, err = s.getApplicationDestination(ctx, q.Name)
-		if err != nil && !apierr.IsNotFound(err) && !q.Force {
-			return nil, err
-		}
+	a, err := s.appclientset.ArgoprojV1alpha1().Applications(s.ns).Get(*q.Name, metav1.GetOptions{})
+	if err != nil && !apierr.IsNotFound(err) {
+		return nil, err
 	}
 
-	if server != "" && namespace != "" {
-		clst, err := s.db.GetCluster(ctx, server)
-		if err != nil && !q.Force {
-			return nil, err
-		}
-		if clst != nil {
-			config := clst.RESTConfig()
-			err = kube.DeleteResourceWithLabel(config, namespace, common.LabelApplicationName, q.Name)
-			if err != nil && !q.Force {
+	if q.Cascade != nil {
+		if *q.Cascade != a.CascadedDeletion() {
+			a.SetCascadedDeletion(true)
```

Comment:
This seems like a bug? If q.Cascade is false, but a.CascadedDeletion is true,  then we always set cascaded deletion to be true and do not honor the request from the API.